app/src/main/java/org/mozilla/rocket/privately/browse/BrowserFragment.kt

```diff
@@ -425,56 +443,57 @@ class BrowserFragment : LocaleAwareFragment(),
+        override fun onSessionAdded(session: Session) {
+            chromeViewModel.onTabCountChanged(sessionManager.size)
         }
 
-        override fun handleExternalUrl(url: String?): Boolean {
-            return fragment.context?.let {
-                IntentUtils.handleExternalUri(it, url)
-            } ?: false
+        override fun onSessionRemoved(session: Session) {
+            session.unregister(sessionObserver)
+            chromeViewModel.onTabCountChanged(sessionManager.size)
         }
 
-        override fun onShowFileChooser(
-            es: TabViewEngineSession,
-            filePathCallback: ValueCallback<Array<Uri>>?,
-            fileChooserParams: WebChromeClient.FileChooserParams?
-        ): Boolean {
-            // do nothing, exist for interface compatibility only.
-            return false
-        }
-
-        var callback: FullscreenCallback? = null
-        var session: Session? = null
-
-        override fun onSessionAdded(session: Session, arguments: Bundle?) {
-        }
+        override fun onSessionSelected(session: Session) {
+            lastSession?.unregister(sessionObserver)
+            session.register(sessionObserver)
+            lastSession = session
 
-        override fun onProgress(session: Session, progress: Int) {
-            fragment.progressView.progress = progress
+            chromeViewModel.run {
+                onFocusedUrlChanged(session.url)
+                onFocusedTitleChanged(session.title)
+                onNavigationStateChanged(session.canGoBack, session.canGoForward)
+            }
         }
+    }
 
-        override fun onTitleChanged(session: Session, title: String?) {
-            fragment.chromeViewModel.onFocusedTitleChanged(title)
-            session.let {
-                if (fragment.displayUrlView.text.toString() != it.url) {
-                    fragment.displayUrlView.text = it.url
+    private val sessionObserver = object : Session.Observer {
+        override fun onDownload(session: Session, download: mozilla.components.browser.session.Download): Boolean {
+            activity.let {
+                if (it == null || !it.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)) {
+                    return false
                 }
             }
-        }
 
-        override fun onLongPress(session: Session, hitTarget: HitTarget) {
-            fragment.activity?.let {
-                WebContextMenu.show(true,
-                        it,
-                        PrivateDownloadCallback(fragment, session.url),
-                        hitTarget)
-            }
+            val d = Download(
+                    download.url,
+                    download.fileName,
+                    download.userAgent,
+                    "",
+                    download.contentType,
+                    download.contentLength ?: 0,
+                    false
+            )
+            permissionHandler.tryAction(
```

Comment:
I can't download files. But this may not relate to the code change here.

Reply:
download should be worked now